core/src/apps/cardano/helpers/hash_builder_collection.py

```diff
@@ -0,0 +1,110 @@
+from apps.common import cbor
+
+if False:
+    from typing import Any, Generic, Generic as GenericKVP, TypeVar
+    from trezor.utils import HashContext
+
+    T = TypeVar("T")
+    K = TypeVar("K")
+    V = TypeVar("V")
+else:
+    T = 0  # type: ignore
+    K = 0  # type: ignore
+    V = 0  # type: ignore
+    Generic = [object]  # type: ignore
+    GenericKVP = {(0, 0): object}  # type: ignore
+
+
+class HashBuilderCollection:
+    def __init__(self, size: int) -> None:
+        self.size = size
+        self.remaining = size
+        self.hash_fn: HashContext | None = None
+        self.parent: "HashBuilderCollection" | None = None
+        self.has_unfinished_child = False
+
+    def start(self, hash_fn: HashContext) -> "HashBuilderCollection":
+        self.hash_fn = hash_fn
+        self.hash_fn.update(self._header_bytes())
+        return self
+
+    def _insert_child(self, child: "HashBuilderCollection") -> None:
+        child.parent = self
+        assert self.hash_fn is not None
+        child.hash_fn = self.hash_fn
+        self.has_unfinished_child = True
+        child.hash_fn.update(child._header_bytes())
+
+    def _do_enter_item(self) -> None:
+        assert self.hash_fn is not None
+        assert self.remaining > 0
+        if self.has_unfinished_child:
+            raise RuntimeError  # can't add item until child is finished
+
+        self.remaining -= 1
+
+    def _hash_item(self, item: Any) -> None:
+        assert self.hash_fn is not None
+        for chunk in cbor.encode_streamed(item):
+            self.hash_fn.update(chunk)
+
+    def _header_bytes(self) -> bytes:
+        raise NotImplementedError
+
+    def finish(self) -> None:
+        if self.remaining != 0:
+            raise RuntimeError  # not all items were added
+        if self.parent is not None:
+            self.parent.has_unfinished_child = False
+        self.hash_fn = None
+        self.parent = None
+
+    def finish_and_get_hash(self) -> bytes:
+        assert self.hash_fn is not None
+        hash = self.hash_fn.digest()
```

Comment:
you should first `self.finish()` and then get digest

the alternative is to avoid this function and do:
```
hash_fn = blake2b(...)
tx_dict.start(hash_fn)
with tx_dict:
	...
digest = hash_fn.digest()
```

Reply:
https://github.com/trezor/trezor-firmware/pull/1683/commits/ed157b3fd72aa0fa789980af3597d201ce21672a